### PR TITLE
feat: show highest permission role when join community

### LINKED ISSUE #18127: Communities: Show highest permission role according to current address selection
# Feature Issue

### User Story

- As a user joining a token-gated community
- When I am selecting addresses for permissions
- I want to see the highest permission role I can join
- So that I can make a more informed decision about which addresses to share.

### Additional requirements

There are 4 places that would show the user's highest permission role:

<table>
  <thead>
    <tr>
      <th>Overview screen</th>
      <th>Request to join screen</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th>Address selection screen</th>
      <th>All permissions drawer</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>

### Suggestions

- The highest permission role should ideally come from status-go. Double-check if this is not already implemented there. We may consider returning this piece of data in the same endpoint (may not exist now on 2023-12-08) that returns the data with all addresses and relevant tokens (implemented as part of https://github.com/status-im/status-mobile/issues/18126)

#### Action Point

Ask desktop team if this functionality (highest permission role should ideally come from status-go) has been implemented.

### Resources

- Figma https://www.figma.com/file/h9wo4GipgZURbqqr1vShFN/Communities-for-Mobile?type=design&node-id=15331-289598&mode=design&t=AIWaz4wOOOzVbMok-0

### Blocked by

- #18126
- status-go work, TODO add issue link once created.

### LINKED ISSUE #18581: Schema error after recent changes on token-tag
token image src might be `nil`

- sub :community/token-gated-overview
- token-requirement-list
- token-tag
- token view

related changes https://github.com/status-im/status-mobile/blob/eeb3c9853bab99c6311cc0cef47c352ee9814326/src/quo/components/tags/token_tag/view.cljs

CleanShot 2024-01-22 at 09 53 43

CleanShot 2024-01-22 at 09 59 03@2x

CleanShot 2024-01-22 at 09 54 19@2x

cc @ulisesmac

### PR DESCRIPTION
**Pull request by yqrashawn**
fixes #18127
fixes #18581

### Summary

feature behind community-accounts-selection-enabled? flag

- [x] community overview screen
- [x] request to join screen
- [x] address selection screen
- [ ] all permissions drawer -> this drawer is not implemented yet
- [x] get revealed accounts (shared addrs) once joined

rely on 
- #18465
- #18549 (picked into this PR)
- #18571

[status-im/status-mobile/assets/15090582/daedf620-7e16-40be-b9f8-3d238afa7d1d](https://github.com/status-im/status-mobile/assets/15090582/daedf620-7e16-40be-b9f8-3d238afa7d1d)

### CODE REVIEW

File: src/status_im/contexts/communities/actions/accounts_selection/view.cljs
Comment (ilmotta) [inline src/status_im/contexts/communities/actions/accounts_selection/view.cljs]: Small refactoring suggestion: extract to a function `role->translation-key`

File: src/status_im/contexts/communities/actions/accounts_selection/view.cljs
Comment (ilmotta) [inline src/status_im/contexts/communities/actions/accounts_selection/view.cljs]: `highest-permission-role-*` is a very clear name. Perhaps shortening it to just `highest-role-*` gives the same clarity.

File: src/status_im/contexts/communities/events.cljs
Comment (ilmotta) [inline src/status_im/contexts/communities/events.cljs]: To follow our current standards, this namespace already contains logic that should be moved to the data store namespaces, particularly the `*-rpc` functions. It would be nice if you could write the new functions already in the data store directory, such as `<-revealed-accounts-rpc`.

File: src/status_im/contexts/communities/events_test.cljs
Comment (ilmotta) [inline src/status_im/contexts/communities/events_test.cljs]: It's a bit hard to reason what these nils represent without having to check the implementation and the order of `when` usages. We could have a margin comment on each one, or maybe better just remove nils and only assert what actually gets dispatched.

File: src/status_im/subs/communities.cljs
Comment (ilmotta) [inline src/status_im/subs/communities.cljs:316]: This subscription is turning out to be somewhat complex to understand. I don't know if it's worth refactoring it too much because we can actually consider having a single endpoint in status-go to return all of this, but since we don't have that yet, consider refactoring it into small private functions in this sub namespace, this would go a long of helping understand the code.

File: src/status_im/contexts/communities/events.cljs
Comment (ilmotta) [inline src/status_im/contexts/communities/events.cljs:396]: 👏🏼 👏🏼

**Review by ilmotta — COMMENTED** [review]
@yqrashawn, since the PR is still in progress, I did a quick review pass. Let me know once you have fully finished it and I can review again. Thank you!

**Comment by yqrashawn** [comment]
@ilmotta no wonder this PR didn't get review, I forgot to remove the WIP in the title.

PR is ready for review

File: src/status_im/contexts/communities/events.cljs
Comment (ilmotta) [inline src/status_im/contexts/communities/events.cljs]: This solution here is probably okay for your PR, but as we know @ajayesivan is also working in a PR that is trying to fix all places that need to consume the sequence of token permissions. Perhaps after you merge this one, or if Ajay merges his one first we should try to use a unified predicate to tell if there are membership permissions, just so we avoid bugs because we might forget to update all places with duplicated logic.

**Review by ilmotta — APPROVED** [review]
Nice work @yqrashawn. There's some room for some refactorings as I suggested, but nothing that can't be done in future PRs related to account selection.

File: src/status_im/subs/communities.cljs
Comment (yqrashawn) [reply src/status_im/subs/communities.cljs:316]: The permissions and permissions-check in db community are in weird shapes (maybe there's a reason?) and do need some refactors on both mobile and go side

> having a single endpoint in status-go to return all of this

the approach using `wakuext_fetchCommunity` to re-fetch community should be removed and rely on status-go signal status-mobile about community updates
I think permission updates should be handled that way as well (already partially handled? since the video shows mobile react and perm change automatically)

`:communities/check-permissions-to-join-community` (to get token-permissions-check) do many requests to blockchain nodes on every call
which might be a problem if merge it into the same endpoint

maybe balance change -> go signals changes to mobile is a better approach
since after the include all future address feature new wallet account will also trigger changes

File: src/status_im/contexts/communities/actions/accounts_selection/view.cljs
Comment (yqrashawn) [reply src/status_im/contexts/communities/actions/accounts_selection/view.cljs]: changed to highest-role-text

File: src/status_im/contexts/communities/actions/accounts_selection/view.cljs
Comment (yqrashawn) [reply src/status_im/contexts/communities/actions/accounts_selection/view.cljs]: added to status-im.contexts.communities.utils

File: src/status_im/contexts/communities/events.cljs
Comment (yqrashawn) [reply src/status_im/contexts/communities/events.cljs]: moved all these <- fn to legacy.status-im.data-store.communities

File: src/status_im/contexts/communities/events_test.cljs
Comment (yqrashawn) [reply src/status_im/contexts/communities/events_test.cljs]: removed with filter some?

File: src/status_im/subs/communities.cljs
Comment (ilmotta) [reply src/status_im/subs/communities.cljs:316]: Good points. I don't know either if the signals are properly handled by us, but looks like it as you said. Something to be checked by us.

I agree with you @yqrashawn, the pull approach to get balances is usually not a good idea. We could cache on the backend, but at the end of the line to join a community, some check(s) against chain(s) must be performed anyway.

The endpoint to provide what I call "permissioned balances" (relevant tokens per Figma) I'm working is also implemented by pulling data from the chains. I haven't looked at a signals approach, but now that you say, I think we should, although I think the machinery is not fully ready in status-go. For instance, the permissions checker in status-go does sync calls to the chains on every check.

**Pull request merged**
